### PR TITLE
chore: Update Badger version to v4

### LINKED ISSUE #1702: TestQueryOneToOneRelations flaky in CI
`TestQueryOneToOneRelations` often fails on the change detector on the CI. It's not clear yet what is causing this.

It looks to be #1672 - only now flaky instead of consistently failing (and on a different, but similar test).

Possibly significant is that both this and #1672 involved tests with one-one relations.

- Executing `TestQueryOneToOneRelations` only does not resolve the issue (other tests do not appear to be a factor)
- Removing the create statements for either `Publishers` or `Authors` appears to remove the issue
- Reducing the query to just `Book {name}` appears to resolve the issue - this is unlike #1672 which persisted on a simplified query
     - Reducing the query to `Book {name author {name}}` (removing the publisher join) also resolves the issue
     - Reducing the query to `Book {name publisher {name}}` (removing the author join) *does not* resolve the issue. This interestingly is the one-one join (same as #1672)
- Setting `n.subtype.Spans()` does not resolve the issue (is a difference between the primary and secondary)
- No errors in `valuesSecondary` appear to be generated, despite #1711
- Returning early from `valuesSecondary` immediately after `n.subType.Init()` resolves the issue, this is before `n.subType.Next()`
- Returning early from `valuesSecondary` immediately after `n.subType.Next()` does not resolve the issue, suggesting `Next` is the trigger (same as in #1672)
- Removing `n.subType.Init()` from `typeJoinOne.Init()` does not resolve the issue (although it is a duplicate, pointless call I think)
- Updating to badger v4 still has no affect
- ~~Removing the 'extra' `df.kvResultsIter.Close()` call added to the fetcher in #1672  appears to resolve the issue~~ It did for a while, and then broke following a git push

### PR DESCRIPTION
**Pull request by fredcarle**
## Relevant issue(s)

Resolves #1702

## Description

This PR updates the Badger version to v4.

It includes a temporary fix to our badger `os.Exit` issue that is costing us time when managing PRs. This should be reverted or properly fixed before releasing v0.7.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

multiple runs of `env DEFRA_DETECT_DATABASE_CHANGES=true gotestsum -- ./tests/integration/query/one_to_many_to_one -shuffle=on -p 1`

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -202,6 +201,7 @@ require (
 
 // SourceNetwork fork og graphql-go
 replace (
+	github.com/dgraph-io/badger/v4 => github.com/fredcarle/badger/v4 v4.0.0-20230801145501-d3a57bd4c2ec
```
Comment (AndrewSisley) [inline go.mod:204]: question: I thought we agreed to use a source fork instead of a personal one? I do have a preference for a sourcenetwork fork.

**Comment by AndrewSisley** [comment]
question: Has the fork been reviewed?

File: go.mod (hunk 1 above)
Comment (fredcarle) [reply go.mod:204]: Used mine because I didn't have the rights to push to our fork. I can update it now that I do have rights but it's temporary anyways.

**Comment by fredcarle** [comment]
> question: Has the fork been reviewed?

Initial review has been done and it's not going to be the way to fix this but the badger team is also looking into it. Hopefully we get a proper fix before the end of this cycle.

File: go.mod (hunk 1 above)
Comment (AndrewSisley) [reply go.mod:204]: Cheers :)

**Review by AndrewSisley — APPROVED** [review]
LGTM, thanks for switching to the source fork :)

**Pull request merged**